File: 41.first-missing-positive/solution.cpp
class Solution {
public:
	int firstMissingPositive(vector<int>& nums) {
		for (int i = 0; i < nums.size(); i++)
		{
			int j = i;
			while (nums[j] < nums.size() && nums[j] > 0 && nums[j] != j + 1 && nums[j] != nums[nums[j] - 1])
			{
				int temp = nums[j];
				nums[j] = nums[temp - 1];
				nums[temp - 1] = temp;
			}
		}
		for (int i = 0; i < nums.size(); i++)
		{
			if (nums[i] != i + 1)
			{
				return i + 1;
			}
		}
		return nums.size() + 1;
	}
};

File: 41.first-missing-positive/solution.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        for i in range(len(nums)):
            while True:
                if nums[i] <= 0 or nums[i] > len(nums) or nums[i] - 1 == i or nums[i] == nums[nums[i] - 1]:
                    break
                tmp = nums[nums[i] - 1]
                nums[nums[i] - 1] = nums[i]
                nums[i] = tmp
        for i in range(len(nums)):
            if i != nums[i] - 1:
                return i + 1
        return len(nums) + 1
